// comfy-tslint/dist/linter.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.lintFile = void 0;
var typescript_1 = require("typescript");
var types_1 = require("./types");
// Determines if this is a test file.
var isTestFile = function (file) {
    return file.fileName.endsWith("_test.ts") ||
        file.fileName.endsWith("_test.tsx");
};
// Determines whether this is a ++var, --var, var++, or var--.
var isVariableIncDec = function (expr) {
    if (expr.kind === typescript_1.SyntaxKind.PrefixUnaryExpression) {
        if ((0, types_1.assertPrefixUnaryExpr)(expr)) {
            return (expr.operator === typescript_1.SyntaxKind.PlusPlusToken ||
                expr.operator === typescript_1.SyntaxKind.MinusMinusToken) &&
                expr.operand.kind === typescript_1.SyntaxKind.Identifier;
        }
    }
    else if (expr.kind === typescript_1.SyntaxKind.PostfixUnaryExpression) {
        if ((0, types_1.assertPostfixUnaryExpr)(expr)) {
            return (expr.operator === typescript_1.SyntaxKind.PlusPlusToken ||
                expr.operator === typescript_1.SyntaxKind.MinusMinusToken) &&
                expr.operand.kind === typescript_1.SyntaxKind.Identifier;
        }
    }
    return false;
};
/**
 * Lints the contents of the given file.
 * @param file File to lint
 */
var lintFile = function (file) {
    file.root.forEachChild(function (node) { return lintTopLevelNode(file, node); });
};
exports.lintFile = lintFile;
// Lints the given child of the source file.
var lintTopLevelNode = function (file, node) {
    switch (node.kind) {
        case typescript_1.SyntaxKind.ImportDeclaration:
            if ((0, types_1.assertImportDecl)(node)) {
                lintImportDecl(file, node);
            }
            break;
        case typescript_1.SyntaxKind.EndOfFileToken:
            // OK
            break;
        case typescript_1.SyntaxKind.TypeAliasDeclaration:
            if ((0, types_1.assertTypeAliasDecl)(node)) {
                lintTypeAliasDecl(file, node);
            }
            break;
        case typescript_1.SyntaxKind.VariableStatement:
            if ((0, types_1.assertVariableStmt)(node)) {
                lintVariableStmt(file, node, true, false);
            }
            break;
        case typescript_1.SyntaxKind.ExpressionStatement:
            if ((0, types_1.assertExpressionStmt)(node)) {
                lintExpressionStmt(file, node, false);
            }
            break;
        case typescript_1.SyntaxKind.IfStatement:
            if ((0, types_1.assertIfStmt)(node)) {
                lintIfStmt(file, node, false);
            }
            break;
        case typescript_1.SyntaxKind.SwitchStatement:
            if ((0, types_1.assertSwitchStmt)(node)) {
                lintSwitchStmt(file, node, false);
            }
            break;
        case typescript_1.SyntaxKind.WhileStatement:
            if ((0, types_1.assertWhileStmt)(node)) {
                lintWhileStmt(file, node, false);
            }
            break;
        case typescript_1.SyntaxKind.ForStatement:
            if ((0, types_1.assertForStmt)(node)) {
                lintForStmt(file, node, false);
            }
            break;
        case typescript_1.SyntaxKind.ForOfStatement:
            if ((0, types_1.assertForOfStmt)(node)) {
                lintForOfStmt(file, node, false);
            }
            break;
        case typescript_1.SyntaxKind.ReturnStatement:
            if ((0, types_1.assertReturnStmt)(node)) {
                lintReturnStmt(file, node, false);
            }
            break;
        case typescript_1.SyntaxKind.ThrowStatement:
            if ((0, types_1.assertThrowStmt)(node)) {
                lintThrowStmt(file, node, false);
            }
            break;
        case typescript_1.SyntaxKind.TryStatement:
            if ((0, types_1.assertTryStmt)(node)) {
                lintTryStmt(file, node, false);
            }
            break;
        case typescript_1.SyntaxKind.DoStatement:
            if ((0, types_1.assertDoStmt)(node)) {
                file.error("do statements are not allowed", node.statement.pos - 2);
            }
            break;
        case typescript_1.SyntaxKind.ForInStatement:
            if ((0, types_1.assertForInStmt)(node)) {
                file.error("for-in statements are not allowed", node.expression.pos, node.expression.end);
            }
            break;
        case typescript_1.SyntaxKind.WithStatement:
            if ((0, types_1.assertWithStmt)(node)) {
                file.error("with statements are not allowed", node.expression.pos, node.expression.end);
            }
            break;
        case typescript_1.SyntaxKind.LabeledStatement:
            if ((0, types_1.assertLabeledStmt)(node)) {
                file.error("labeled statements are not allowed", node.label.end - 1);
            }
            break;
        case typescript_1.SyntaxKind.Block:
            if ((0, types_1.assertBlock)(node)) {
                lintBlock(file, node, false);
            }
            break;
        case typescript_1.SyntaxKind.FunctionDeclaration:
            if ((0, types_1.assertFunctionDecl)(node)) {
                if (node.name === undefined) {
                    file.error("incorrect syntax for a function declaration", node.parameters.pos);
                }
                else {
                    file.error("incorrect syntax for a function declaration", node.name.pos);
                }
            }
            break;
        case typescript_1.SyntaxKind.InterfaceDeclaration:
            if ((0, types_1.assertInterfaceDecl)(node)) {
                lintInterfaceDecl(file, node);
            }
            break;
        case typescript_1.SyntaxKind.ClassDeclaration:
            if ((0, types_1.assertClassDecl)(node)) {
                lintClassDecl(file, node);
            }
            break;
        case typescript_1.SyntaxKind.ModuleDeclaration:
            // Used in saferoutes.ts. We allow it to go through
            break;
        default:
            file.error("unexpected top-level node type: ".concat(typescript_1.SyntaxKind[node.kind]), node.pos + 1);
            break;
    }
};
// Set of imports that can be used with "import name as 'name'" syntax.
var allowedNameImports = new Set(["express", "body-parser", "react"]);
// Set of imports that can be used with "import * from name as 'name'" syntax.
var allowedStarImports = new Set(["assert", "node-mocks-http"]);
/** Determines if the given name could be an image file. */
var isImageName = function (name) {
    name = name.toLowerCase();
    return name.endsWith(".png") || name.endsWith(".jpg") || name.endsWith(".jpeg") ||
        name.endsWith(".gif") || name.endsWith(".svg");
};
// Performs linting of an import statement.
var lintImportDecl = function (file, node) {
    if (node.importClause === undefined) {
        // OK (this is probably a CSS import)
    }
    else {
        if (node.importClause.namedBindings !== undefined &&
            node.importClause.namedBindings.kind === typescript_1.SyntaxKind.NamedImports) {
            // OK (regular import)
        }
        else if (node.importClause.name !== undefined &&
            node.importClause.name.kind === typescript_1.SyntaxKind.Identifier &&
            node.moduleSpecifier !== undefined &&
            node.moduleSpecifier.kind === typescript_1.SyntaxKind.StringLiteral) {
            if ((0, types_1.assertStringLiteral)(node.moduleSpecifier)) {
                if (!allowedNameImports.has(node.moduleSpecifier.text) &&
                    !isImageName(node.moduleSpecifier.text)) {
                    file.error("unexpected import of ".concat(node.moduleSpecifier.text), node.importClause.pos);
                }
            }
        }
        else if (node.importClause.namedBindings !== undefined &&
            node.importClause.namedBindings.kind === typescript_1.SyntaxKind.NamespaceImport &&
            node.moduleSpecifier !== undefined &&
            node.moduleSpecifier.kind === typescript_1.SyntaxKind.StringLiteral) {
            if ((0, types_1.assertStringLiteral)(node.moduleSpecifier)) {
                if (!allowedStarImports.has(node.moduleSpecifier.text)) {
                    file.error("unexpected * import of ".concat(node.moduleSpecifier.text), node.importClause.pos);
                }
            }
        }
        else {
            file.error("unexpected import syntax", node.importClause.pos);
        }
    }
};
// Performs linting of a type alias declaration.
var lintTypeAliasDecl = function (file, node) {
    lintType(file, node.type, true);
};
// Performs linting of an interface declaration.
var lintInterfaceDecl = function (file, node) {
    var e_1, _a, e_2, _b;
    try {
        for (var _c = __values(node.members), _d = _c.next(); !_d.done; _d = _c.next()) {
            var member = _d.value;
            if (member.kind === typescript_1.SyntaxKind.PropertySignature) {
                if ((0, types_1.assertPropertySignature)(member)) {
                    if (member.type === undefined) {
                        file.error("field declaration is missing a type", member.name.end - 1);
                    }
                    else {
                        if (member.type.kind !== typescript_1.SyntaxKind.FunctionType) {
                            file.error("interfaces can only contain methods", member.name.end - 1);
                        }
                        else {
                            if ((0, types_1.assertFunctionType)(member.type)) {
                                try {
                                    for (var _e = (e_2 = void 0, __values(member.type.parameters)), _f = _e.next(); !_f.done; _f = _e.next()) {
                                        var param = _f.value;
                                        if (param.type === undefined) {
                                            file.error("parameter is missing a type", param.name.end - 1);
                                        }
                                        else {
                                            lintType(file, param.type);
                                        }
                                    }
                                }
                                catch (e_2_1) { e_2 = { error: e_2_1 }; }
                                finally {
                                    try {
                                        if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
                                    }
                                    finally { if (e_2) throw e_2.error; }
                                }
                                lintType(file, member.type.type);
                            }
                        }
                    }
                }
            }
            else if (member.kind === typescript_1.SyntaxKind.MethodSignature) {
                if ((0, types_1.assertMethodSignature)(member)) {
                    file.error("incorrect method declaration syntax", member.name.end - 1);
                }
            }
            else {
                file.error("unexpected interface member type: ".concat(typescript_1.SyntaxKind[member.kind]), member.pos + 1);
            }
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
        }
        finally { if (e_1) throw e_1.error; }
    }
};
// Performs linting of an class declaration.
var lintClassDecl = function (file, node) {
    var e_3, _a;
    var superclass = undefined;
    var interfaces = [];
    if (node.heritageClauses !== undefined) {
        try {
            for (var _b = __values(node.heritageClauses), _c = _b.next(); !_c.done; _c = _b.next()) {
                var herit = _c.value;
                if (herit.types.length === 1 &&
                    herit.types[0].expression.kind === typescript_1.SyntaxKind.Identifier &&
                    (0, types_1.assertIdentifier)(herit.types[0].expression)) {
                    if (herit.token === typescript_1.SyntaxKind.ExtendsKeyword) {
                        superclass = herit.types[0].expression.text;
                    }
                    else if (herit.token === typescript_1.SyntaxKind.ImplementsKeyword) {
                        interfaces.push(herit.types[0].expression.text);
                    }
                    else {
                        file.error("unexpected inheritance token: ".concat(typescript_1.SyntaxKind[herit.token]), herit.pos + 1);
                    }
                }
                else {
                    file.error("incorrect inheritance syntax", herit.pos + 1);
                }
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_3) throw e_3.error; }
        }
    }
    if (superclass === "Component") {
        lintComponentClass(file, node, interfaces);
    }
    else {
        lintAdtClass(file, node, superclass, interfaces);
    }
};
// Performs linting of a UI component.
var lintComponentClass = function (file, node, interfaces) {
    var e_4, _a;
    if (interfaces.length > 0) {
        file.error("components should not implement any interfaces", node.name !== undefined ? node.name.pos : node.pos + 1);
    }
    try {
        for (var _b = __values(node.members), _c = _b.next(); !_c.done; _c = _b.next()) {
            var member = _c.value;
            switch (member.kind) {
                case typescript_1.SyntaxKind.Constructor:
                    if ((0, types_1.assertConstructorDecl)(member)) {
                        lintConstructor(file, member);
                    }
                    break;
                case typescript_1.SyntaxKind.PropertyDeclaration:
                    if ((0, types_1.assertPropertyDecl)(member)) {
                        if (member.initializer !== undefined &&
                            member.initializer.kind === typescript_1.SyntaxKind.ArrowFunction &&
                            (0, types_1.assertArrowFunction)(member.initializer)) {
                            var name_1 = getName(member.name);
                            if (name_1 !== undefined) {
                                lintComponentMethodName(file, member.name, name_1);
                            }
                            else {
                                file.error("unexpected property name syntax: ".concat(typescript_1.SyntaxKind[member.name.kind]), member.name.end);
                            }
                            lintArrowFunction(file, member.initializer, false);
                        }
                        else {
                            file.error("fields are not allowed in components", member.name.end - 1);
                        }
                    }
                    break;
                case typescript_1.SyntaxKind.MethodDeclaration:
                    if ((0, types_1.assertMethodDecl)(member)) {
                        file.error("incorrect method declaration syntax", member.name.end - 1);
                    }
                    break;
                default:
                    file.error("unexpected class member type: ".concat(typescript_1.SyntaxKind[member.kind]), member.pos + 1);
                    break;
            }
        }
    }
    catch (e_4_1) { e_4 = { error: e_4_1 }; }
    finally {
        try {
            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
        }
        finally { if (e_4) throw e_4.error; }
    }
};
// Set of allowed lifecycle methods.
var allowedLifeCycleMethods = new Set(["componentDidMount", "componentDidUpdate", "componentWillUnmount"]);
// Set of allowed HTML events.
// TODO: add more
var allowedEvents = new Set([
    "Click", "Change", "Timeout", "Focus", "Blur", "MouseEnter", "MouseLeave",
    "KeyDown", "KeyUp", "KeyPress", "Fetch", "Resp", "Response", "Json",
    "Error"
]);
// Performs linting of the name of a component method
var lintComponentMethodName = function (file, node, name) {
    var e_5, _a;
    if (name.startsWith("render") &&
        (name.length === 6 || name[6].toUpperCase() === name[6])) {
        // OK: render or render helper
    }
    else if (name.startsWith("get") &&
        name.length > 3 && name[3].toUpperCase() === name[3]) {
        // OK: getter function
    }
    else if (name.startsWith("set") &&
        name.length > 3 && name[3].toUpperCase() === name[3]) {
        // OK: setter function
    }
    else if (allowedLifeCycleMethods.has(name)) {
        // OK
    }
    else {
        if (!name.startsWith("do") || name.length < 3 ||
            name[2].toUpperCase() !== name[2]) {
            file.error("event handler method names must start with 'do'", node.end - 1);
        }
        else {
            try {
                for (var allowedEvents_1 = __values(allowedEvents), allowedEvents_1_1 = allowedEvents_1.next(); !allowedEvents_1_1.done; allowedEvents_1_1 = allowedEvents_1.next()) {
                    var evtName = allowedEvents_1_1.value;
                    if (name.endsWith(evtName)) {
                        return; // OK
                    }
                }
            }
            catch (e_5_1) { e_5 = { error: e_5_1 }; }
            finally {
                try {
                    if (allowedEvents_1_1 && !allowedEvents_1_1.done && (_a = allowedEvents_1.return)) _a.call(allowedEvents_1);
                }
                finally { if (e_5) throw e_5.error; }
            }
            file.error("event handler method names must end with an event name (e.g., Click)", node.end);
        }
    }
};
// Performs linting of a class that implements an ADT.
var lintAdtClass = function (file, node, superclass, interfaces) {
    var e_6, _a;
    if (node.modifiers !== undefined &&
        contains(node.modifiers, typescript_1.SyntaxKind.ExportKeyword)) {
        file.error("classes cannot be exported", node.modifiers.pos);
    }
    if (superclass !== undefined) {
        file.error("inheritance is not allowed", node.name !== undefined ? node.name.end - 1 : node.pos + 1);
    }
    if (interfaces.length === 0) {
        file.error("classes must implement some interface", node.name !== undefined ? node.name.end - 1 : node.pos + 1);
    }
    try {
        for (var _b = __values(node.members), _c = _b.next(); !_c.done; _c = _b.next()) {
            var member = _c.value;
            switch (member.kind) {
                case typescript_1.SyntaxKind.Constructor:
                    if ((0, types_1.assertConstructorDecl)(member)) {
                        lintConstructor(file, member);
                    }
                    break;
                case typescript_1.SyntaxKind.PropertyDeclaration:
                    if ((0, types_1.assertPropertyDecl)(member)) {
                        if (member.initializer === undefined) {
                            if (member.type === undefined) {
                                file.error("field is missing a type", member.name.end - 1);
                            }
                            else {
                                lintType(file, member.type);
                            }
                        }
                        else {
                            if (member.initializer.kind === typescript_1.SyntaxKind.ArrowFunction &&
                                (0, types_1.assertArrowFunction)(member.initializer)) {
                                lintArrowFunction(file, member.initializer, false);
                            }
                            else {
                                lintExpression(file, member.initializer, false);
                            }
                        }
                    }
                    break;
                case typescript_1.SyntaxKind.MethodDeclaration:
                    if ((0, types_1.assertMethodDecl)(member)) {
                        file.error("incorrect method declaration syntax", member.name.end - 1);
                    }
                    break;
                default:
                    file.error("unexpected class member type: ".concat(typescript_1.SyntaxKind[member.kind]), member.pos + 1);
                    break;
            }
        }
    }
    catch (e_6_1) { e_6 = { error: e_6_1 }; }
    finally {
        try {
            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
        }
        finally { if (e_6) throw e_6.error; }
    }
};
// Performs linting of a constructor declaration.
var lintConstructor = function (file, node) {
    var e_7, _a;
    try {
        for (var _b = __values(node.parameters), _c = _b.next(); !_c.done; _c = _b.next()) {
            var param = _c.value;
            if (param.type === undefined) {
                file.error("parameters must have types", param.name.end - 1);
            }
            else {
                lintType(file, param.type);
            }
        }
    }
    catch (e_7_1) { e_7 = { error: e_7_1 }; }
    finally {
        try {
            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
        }
        finally { if (e_7) throw e_7.error; }
    }
    if (node.body === undefined) {
        file.error("constructor is missing a body", node.name !== undefined ? node.name.end - 1 : node.pos + 1);
    }
    else {
        lintBlock(file, node.body, true);
    }
};
// Performs linting of a *top-level* variable statement.
var lintVariableStmt = function (file, node, topLevel, isConstructor) {
    if (node.modifiers !== undefined &&
        contains(node.modifiers, typescript_1.SyntaxKind.ExportKeyword) &&
        !(0, types_1.hasJSDoc)(node)) {
        file.error("exported declarations must have a JSDoc comment", (node.declarationList.declarations.length === 0) ? node.pos + 1 :
            node.declarationList.declarations[0].name.end - 1);
    }
    lintVariableDecl(file, node.declarationList, topLevel, false, isConstructor);
};
// Performs linting of a variable declaration.
var lintVariableDecl = function (file, list, topLevel, reqInitalizer, isConstructor) {
    if (!file.allowMutation && (list.flags & typescript_1.NodeFlags.Const) === 0) {
        file.error("all variable declarations must be const (no mutation)", (list.declarations.length === 0) ? list.pos + 1 :
            list.declarations[0].name.end - 1);
    }
    if ((list.flags & typescript_1.NodeFlags.Const) === 0 && (list.flags & typescript_1.NodeFlags.Let) === 0) {
        file.error("variables can only be declared with let or const (no var)", (list.declarations.length === 0) ? list.pos + 1 :
            list.declarations[0].name.end - 1);
    }
    if (list.declarations.length !== 1) {
        file.error("expected exactly one variable declaration", list.declarations.pos + 1);
    }
    else {
        var decl = list.declarations[0];
        lintBindingName(file, decl.name);
        if (decl.initializer === undefined) {
            if (reqInitalizer) {
                file.error("must be initialized", decl.name.end - 1);
            }
        }
        else if (topLevel && decl.initializer.kind === typescript_1.SyntaxKind.ArrowFunction) {
            if ((0, types_1.assertArrowFunction)(decl.initializer)) {
                lintArrowFunction(file, decl.initializer, isConstructor);
            }
        }
        else {
            if (decl.type === undefined) {
                if (topLevel) {
                    file.error("top-level variable declarations must have a type", decl.name.end - 1);
                }
            }
            else {
                lintType(file, decl.type);
            }
            lintExpression(file, decl.initializer, isConstructor);
        }
    }
};
// Performs linting of a binding name.
var lintBindingName = function (file, name) {
    var e_8, _a;
    if (name.kind === typescript_1.SyntaxKind.Identifier) {
        // OK
    }
    else if (name.kind === typescript_1.SyntaxKind.ArrayBindingPattern) {
        if ((0, types_1.assertArrayBindingPattern)(name)) {
            try {
                for (var _b = __values(name.elements), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var elem = _c.value;
                    if (elem.kind === typescript_1.SyntaxKind.BindingElement) {
                        if ((0, types_1.assertBindingElement)(elem)) {
                            lintBindingName(file, elem.name);
                        }
                    }
                }
            }
            catch (e_8_1) { e_8 = { error: e_8_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_8) throw e_8.error; }
            }
        }
    }
    else { // object binding pattern
        file.error("record destructuring is not allowed", name.end - 1);
    }
};
// Determines whether the given array of modifiers contains the kind.
var contains = function (mods, kind) {
    for (var i = 0; i < mods.length; i++) {
        if (mods[i].kind === kind) {
            return true;
        }
    }
    return false;
};
// Performs linting of an arrow function.
var lintArrowFunction = function (file, node, isConstructor) {
    var e_9, _a;
    var paramTypes = [];
    try {
        for (var _b = __values(node.parameters), _c = _b.next(); !_c.done; _c = _b.next()) {
            var param = _c.value;
            if (param.type === undefined) {
                file.error("parameters must have types", param.name.end - 1);
            }
            else if (param.initializer !== undefined) {
                file.error("parameters cannot have default values", param.name.end - 1);
            }
            else {
                lintType(file, param.type);
                paramTypes.push(param.type);
            }
        }
    }
    catch (e_9_1) { e_9 = { error: e_9_1 }; }
    finally {
        try {
            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
        }
        finally { if (e_9) throw e_9.error; }
    }
    if (node.type === undefined) {
        file.error("functions must have a return type", node.parameters.end + 1);
    }
    else {
        lintType(file, node.type);
    }
    // For .ts files: if the two argument types are (Safe)Request and
    // (Safe)Response, then make sure they are both declared as the safe variants.
    if (paramTypes.length === 2) {
        var paramType0 = paramTypes[0];
        var paramType1 = paramTypes[1];
        if (paramType0.kind === typescript_1.SyntaxKind.TypeReference &&
            paramType1.kind === typescript_1.SyntaxKind.TypeReference) {
            if ((0, types_1.assertTypeReference)(paramType0) &&
                (0, types_1.assertTypeReference)(paramType1)) {
                if (paramType0.typeName.kind === typescript_1.SyntaxKind.Identifier &&
                    paramType1.typeName.kind === typescript_1.SyntaxKind.Identifier) {
                    if ((paramType0.typeName.text === "Request" ||
                        paramType0.typeName.text === "SafeRequest") &&
                        (paramType1.typeName.text === "Response" ||
                            paramType1.typeName.text === "SafeResponse") &&
                        file.fileName.endsWith(".ts")) {
                        if (paramType0.typeName.text !== "SafeRequest") {
                            file.error("must use SafeRequest and SafeResponse", paramType0.typeName.end - 1);
                        }
                        if (paramType1.typeName.text !== "SafeResponse") {
                            file.error("must use SafeRequest and SafeResponse", paramType1.typeName.end - 1);
                        }
                    }
                }
            }
        }
    }
    if (node.body.kind === typescript_1.SyntaxKind.Block) {
        if ((0, types_1.assertBlock)(node.body)) {
            lintBlock(file, node.body, isConstructor);
        }
    }
    else {
        lintExpression(file, node.body, isConstructor);
    }
};
// Performs lintinig of a block.
var lintBlock = function (file, node, isConstructor) {
    var e_10, _a;
    try {
        for (var _b = __values(node.statements), _c = _b.next(); !_c.done; _c = _b.next()) {
            var stmt = _c.value;
            lintStatement(file, stmt, isConstructor);
        }
    }
    catch (e_10_1) { e_10 = { error: e_10_1 }; }
    finally {
        try {
            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
        }
        finally { if (e_10) throw e_10.error; }
    }
};
// Lints the given statement inside a function.
var lintStatement = function (file, node, isConstructor) {
    switch (node.kind) {
        case typescript_1.SyntaxKind.EmptyStatement:
        case typescript_1.SyntaxKind.DebuggerStatement:
        case typescript_1.SyntaxKind.ContinueStatement:
        case typescript_1.SyntaxKind.BreakStatement:
            break;
        case typescript_1.SyntaxKind.VariableStatement:
            if ((0, types_1.assertVariableStmt)(node)) {
                lintVariableStmt(file, node, false, isConstructor);
            }
            break;
        case typescript_1.SyntaxKind.ExpressionStatement:
            if ((0, types_1.assertExpressionStmt)(node)) {
                lintExpressionStmt(file, node, isConstructor);
            }
            break;
        case typescript_1.SyntaxKind.IfStatement:
            if ((0, types_1.assertIfStmt)(node)) {
                lintIfStmt(file, node, isConstructor);
            }
            break;
        case typescript_1.SyntaxKind.SwitchStatement:
            if ((0, types_1.assertSwitchStmt)(node)) {
                lintSwitchStmt(file, node, isConstructor);
            }
            break;
        case typescript_1.SyntaxKind.WhileStatement:
            if ((0, types_1.assertWhileStmt)(node)) {
                lintWhileStmt(file, node, isConstructor);
            }
            break;
        case typescript_1.SyntaxKind.ForStatement:
            if ((0, types_1.assertForStmt)(node)) {
                lintForStmt(file, node, isConstructor);
            }
            break;
        case typescript_1.SyntaxKind.ForOfStatement:
            if ((0, types_1.assertForOfStmt)(node)) {
                lintForOfStmt(file, node, isConstructor);
            }
            break;
        case typescript_1.SyntaxKind.ReturnStatement:
            if ((0, types_1.assertReturnStmt)(node)) {
                lintReturnStmt(file, node, isConstructor);
            }
            break;
        case typescript_1.SyntaxKind.ThrowStatement:
            if ((0, types_1.assertThrowStmt)(node)) {
                lintThrowStmt(file, node, isConstructor);
            }
            break;
        case typescript_1.SyntaxKind.TryStatement:
            if ((0, types_1.assertTryStmt)(node)) {
                lintTryStmt(file, node, isConstructor);
            }
            break;
        case typescript_1.SyntaxKind.DoStatement:
            if ((0, types_1.assertDoStmt)(node)) {
                file.error("do statements are not allowed", node.statement.pos - 2);
            }
            break;
        case typescript_1.SyntaxKind.ForInStatement:
            if ((0, types_1.assertForInStmt)(node)) {
                file.error("for-in statements are not allowed", node.expression.pos, node.expression.end);
            }
            break;
        case typescript_1.SyntaxKind.WithStatement:
            if ((0, types_1.assertWithStmt)(node)) {
                file.error("with statements are not allowed", node.expression.pos, node.expression.end);
            }
            break;
        case typescript_1.SyntaxKind.LabeledStatement:
            if ((0, types_1.assertLabeledStmt)(node)) {
                file.error("labeled statements are not allowed", node.label.end - 1);
            }
            break;
        case typescript_1.SyntaxKind.Block:
            if ((0, types_1.assertBlock)(node)) {
                lintBlock(file, node, false);
            }
            break;
        default:
            file.error("unexpected statement type: ".concat(typescript_1.SyntaxKind[node.kind]), node.pos + 1);
    }
};
// Lints the given if statement.
var lintIfStmt = function (file, node, isConstructor) {
    lintExpression(file, node.expression, isConstructor);
    lintStatement(file, node.thenStatement, isConstructor);
    if (node.elseStatement !== undefined) {
        lintStatement(file, node.elseStatement, isConstructor);
    }
};
// Lints the given switch statement.
var lintSwitchStmt = function (file, node, isConstructor) {
    var e_11, _a, e_12, _b;
    lintExpression(file, node.expression, isConstructor);
    try {
        for (var _c = __values(node.caseBlock.clauses), _d = _c.next(); !_d.done; _d = _c.next()) {
            var clause = _d.value;
            if (clause.kind === typescript_1.SyntaxKind.CaseClause) {
                lintExpression(file, clause.expression, isConstructor);
            }
            try {
                for (var _e = (e_12 = void 0, __values(clause.statements)), _f = _e.next(); !_f.done; _f = _e.next()) {
                    var stmt = _f.value;
                    lintStatement(file, stmt, isConstructor);
                }
            }
            catch (e_12_1) { e_12 = { error: e_12_1 }; }
            finally {
                try {
                    if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
                }
                finally { if (e_12) throw e_12.error; }
            }
        }
    }
    catch (e_11_1) { e_11 = { error: e_11_1 }; }
    finally {
        try {
            if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
        }
        finally { if (e_11) throw e_11.error; }
    }
};
var INV_RE = /^(\s+|\s*\/\/[^\n]*\n)*\/\/\s+Inv:\s+/;
// Lints the given while statement.
var lintWhileStmt = function (file, node, isConstructor) {
    if (file.reqInvariants) {
        var text = node.getFullText(file.root);
        if (text.match(INV_RE) === null) {
            var index = text.indexOf("while"); // not perfect, but better than nothing
            file.error("while loops must have an invariant", node.pos + index);
        }
    }
    lintExpression(file, node.expression, isConstructor);
    lintStatement(file, node.statement, isConstructor);
};
// Lints the given for statement.
var lintForStmt = function (file, node, isConstructor) {
    if (file.reqInvariants) {
        var text = node.getFullText(file.root);
        var index = text.indexOf("for"); // not perfect, but better than nothing
        file.error("for loops are not allowed (only while and for-each)", node.pos + index);
    }
    if (node.initializer !== undefined) {
        if (node.initializer.kind === typescript_1.SyntaxKind.VariableDeclarationList) {
            if ((0, types_1.assertVariableDeclList)(node.initializer)) {
                lintVariableDecl(file, node.initializer, false, true, isConstructor);
            }
        }
        else {
            lintExpression(file, node.initializer, isConstructor);
        }
    }
    if (node.condition !== undefined) {
        lintExpression(file, node.condition, isConstructor);
    }
    if (node.incrementor !== undefined) {
        // This is the one place that we allow ++ and --. Must be of an identifier
        if (isVariableIncDec(node.incrementor)) {
            // OK
        }
        else {
            lintExpression(file, node.incrementor, isConstructor);
        }
    }
    lintStatement(file, node.statement, isConstructor);
};
// Lints the given for-of statement.
var lintForOfStmt = function (file, node, isConstructor) {
    // NOTE: no invariant ever required
    if (node.initializer.kind === typescript_1.SyntaxKind.VariableDeclarationList) {
        if ((0, types_1.assertVariableDeclList)(node.initializer)) {
            lintVariableDecl(file, node.initializer, false, false, isConstructor);
        }
    }
    else {
        lintExpression(file, node.initializer, isConstructor);
    }
    lintExpression(file, node.expression, isConstructor);
    lintStatement(file, node.statement, isConstructor);
};
// Lints the given return statement.
var lintReturnStmt = function (file, node, isConstructor) {
    if (node.expression !== undefined) {
        lintExpression(file, node.expression, isConstructor);
    }
};
// Lints the given throw statement.
var lintThrowStmt = function (file, node, isConstructor) {
    if (node.expression !== undefined) {
        lintExpression(file, node.expression, isConstructor);
    }
};
// Lints the given try statement.
var lintTryStmt = function (file, node, isConstructor) {
    lintBlock(file, node.tryBlock, isConstructor);
    if (node.catchClause !== undefined) {
        var decl = node.catchClause.variableDeclaration;
        if (decl !== undefined) {
            if (decl.type === undefined) {
                file.error("exception declarations must have a type", decl.name.end - 1);
            }
            else {
                lintType(file, decl.type);
            }
        }
        lintBlock(file, node.catchClause.block, isConstructor);
    }
    if (node.finallyBlock !== undefined) {
        lintBlock(file, node.finallyBlock, isConstructor);
    }
};
// Performs linting of a type.
var lintType = function (file, node, topLevel) {
    var e_13, _a, e_14, _b, e_15, _c, e_16, _d, e_17, _e;
    switch (node.kind) {
        case typescript_1.SyntaxKind.NumberKeyword:
        case typescript_1.SyntaxKind.StringKeyword:
        case typescript_1.SyntaxKind.BooleanKeyword:
        case typescript_1.SyntaxKind.UndefinedKeyword:
        case typescript_1.SyntaxKind.NullKeyword:
        case typescript_1.SyntaxKind.VoidKeyword:
        case typescript_1.SyntaxKind.UnknownKeyword:
        case typescript_1.SyntaxKind.BigIntKeyword:
            // OK
            break;
        case typescript_1.SyntaxKind.LiteralType: // string or integer literal
            // OK
            break;
        case typescript_1.SyntaxKind.AnyKeyword:
            file.error("any type is not allowed", node.pos);
            break;
        case typescript_1.SyntaxKind.TypeReference: // use of a type alias
            if ((0, types_1.assertTypeReference)(node)) {
                if (node.typeArguments !== undefined) {
                    try {
                        for (var _f = __values(node.typeArguments), _g = _f.next(); !_g.done; _g = _f.next()) { // type arguments
                            var arg = _g.value;
                            lintType(file, arg);
                        }
                    }
                    catch (e_13_1) { e_13 = { error: e_13_1 }; }
                    finally {
                        try {
                            if (_g && !_g.done && (_a = _f.return)) _a.call(_f);
                        }
                        finally { if (e_13) throw e_13.error; }
                    }
                }
            }
            break;
        case typescript_1.SyntaxKind.TypeOperator:
            if ((0, types_1.assertTypeOperator)(node)) {
                lintType(file, node.type);
            }
            break;
        case typescript_1.SyntaxKind.TupleType:
            if ((0, types_1.assertTupleType)(node)) {
                for (var i = 0; i < node.elements.length; i++) {
                    lintType(file, node.elements[i]);
                }
            }
            break;
        case typescript_1.SyntaxKind.UnionType:
            if ((0, types_1.assertUnionType)(node)) {
                try {
                    for (var _h = __values(node.types), _j = _h.next(); !_j.done; _j = _h.next()) {
                        var typ = _j.value;
                        lintType(file, typ);
                    }
                }
                catch (e_14_1) { e_14 = { error: e_14_1 }; }
                finally {
                    try {
                        if (_j && !_j.done && (_b = _h.return)) _b.call(_h);
                    }
                    finally { if (e_14) throw e_14.error; }
                }
            }
            break;
        case typescript_1.SyntaxKind.ArrayType:
            if ((0, types_1.assertArrayType)(node)) {
                lintType(file, node.elementType);
            }
            break;
        case typescript_1.SyntaxKind.TypeLiteral: // record type
            if ((0, types_1.assertTypeLiteral)(node)) {
                try {
                    for (var _k = __values(node.members), _l = _k.next(); !_l.done; _l = _k.next()) {
                        var member = _l.value;
                        if (member.kind === typescript_1.SyntaxKind.PropertySignature) {
                            if ((0, types_1.assertPropertySignature)(member)) {
                                if (member.type === undefined) {
                                    file.error("fields of records must have a type", member.name.end - 1);
                                }
                                else {
                                    lintType(file, member.type);
                                }
                            }
                        }
                        else {
                            file.error("unexpected type literal member: ".concat(typescript_1.SyntaxKind[member.kind]), member.pos + 1);
                        }
                    }
                }
                catch (e_15_1) { e_15 = { error: e_15_1 }; }
                finally {
                    try {
                        if (_l && !_l.done && (_c = _k.return)) _c.call(_k);
                    }
                    finally { if (e_15) throw e_15.error; }
                }
            }
            break;
        case typescript_1.SyntaxKind.FunctionType:
            if ((0, types_1.assertFunctionType)(node)) {
                try {
                    for (var _m = __values(node.parameters), _o = _m.next(); !_o.done; _o = _m.next()) {
                        var param = _o.value;
                        if (param.type === undefined) {
                            file.error("parameters of functions must have a type", param.name.end - 1);
                        }
                        else {
                            lintType(file, param.type);
                        }
                    }
                }
                catch (e_16_1) { e_16 = { error: e_16_1 }; }
                finally {
                    try {
                        if (_o && !_o.done && (_d = _m.return)) _d.call(_m);
                    }
                    finally { if (e_16) throw e_16.error; }
                }
                lintType(file, node.type);
            }
            break;
        case typescript_1.SyntaxKind.TypePredicate:
            if ((0, types_1.assertTypePredicate)(node)) {
                if (node.type !== undefined) {
                    lintType(file, node.type);
                }
            }
            break;
        case typescript_1.SyntaxKind.IntersectionType:
            if ((0, types_1.assertIntersectionType)(node)) {
                if (!topLevel) {
                    file.error("intersection types are only allowed in top-level alias declarations", node.end - 1);
                }
                else {
                    try {
                        for (var _p = __values(node.types), _q = _p.next(); !_q.done; _q = _p.next()) {
                            var typ = _q.value;
                            lintType(file, typ);
                        }
                    }
                    catch (e_17_1) { e_17 = { error: e_17_1 }; }
                    finally {
                        try {
                            if (_q && !_q.done && (_e = _p.return)) _e.call(_p);
                        }
                        finally { if (e_17) throw e_17.error; }
                    }
                }
            }
            break;
        default:
            file.error("unexpected type: ".concat(typescript_1.SyntaxKind[node.kind]), node.pos + 1);
            break;
    }
};
// Performs linting of an expression statement.
var lintExpressionStmt = function (file, node, isConstructor) {
    lintExpression(file, node.expression, isConstructor);
};
// Performs linting of an expression.
var lintExpression = function (file, node, isConstructor) {
    var e_18, _a, e_19, _b, e_20, _c, e_21, _d, e_22, _e;
    switch (node.kind) {
        case typescript_1.SyntaxKind.Identifier:
        case typescript_1.SyntaxKind.NumericLiteral:
        case typescript_1.SyntaxKind.StringLiteral:
        case typescript_1.SyntaxKind.TrueKeyword:
        case typescript_1.SyntaxKind.FalseKeyword:
        case typescript_1.SyntaxKind.NullKeyword:
        case typescript_1.SyntaxKind.ThisKeyword:
        case typescript_1.SyntaxKind.SuperKeyword:
        case typescript_1.SyntaxKind.BigIntLiteral:
            // OK
            break;
        case typescript_1.SyntaxKind.ParenthesizedExpression:
            if ((0, types_1.assertParenthesizedExpr)(node)) {
                lintExpression(file, node.expression, isConstructor);
            }
            break;
        case typescript_1.SyntaxKind.PrefixUnaryExpression:
            if ((0, types_1.assertPrefixUnaryExpr)(node)) {
                if (node.operator === typescript_1.SyntaxKind.PlusPlusToken || node.operator === typescript_1.SyntaxKind.MinusMinusToken) {
                    file.error("++ and -- are not allowed", node.operand.pos - 2);
                }
                lintExpression(file, node.operand, isConstructor);
            }
            break;
        case typescript_1.SyntaxKind.PostfixUnaryExpression:
            if ((0, types_1.assertPostfixUnaryExpr)(node)) {
                file.error("++ and -- are not allowed", node.operand.end + 2);
            }
            break;
        case typescript_1.SyntaxKind.BinaryExpression:
            if ((0, types_1.assertBinaryExpr)(node)) {
                lintBinaryExpr(file, node, isConstructor);
            }
            break;
        case typescript_1.SyntaxKind.ConditionalExpression:
            if ((0, types_1.assertConditionalExpr)(node)) {
                lintExpression(file, node.condition, isConstructor);
                lintExpression(file, node.whenTrue, isConstructor);
                lintExpression(file, node.whenFalse, isConstructor);
            }
            break;
        case typescript_1.SyntaxKind.TypeOfExpression:
            if ((0, types_1.assertTypeOfExpr)(node)) {
                lintExpression(file, node.expression, isConstructor);
            }
            break;
        case typescript_1.SyntaxKind.PropertyAccessExpression:
            if ((0, types_1.assertPropertyAccessExpr)(node)) {
                lintExpression(file, node.expression, isConstructor);
            }
            break;
        case typescript_1.SyntaxKind.ElementAccessExpression:
            if ((0, types_1.assertElementAccessExpr)(node)) {
                lintExpression(file, node.argumentExpression, isConstructor);
                lintExpression(file, node.expression, isConstructor);
            }
            break;
        case typescript_1.SyntaxKind.NewExpression:
            if ((0, types_1.assertNewExpr)(node)) {
                if (node.expression.kind !== typescript_1.SyntaxKind.Identifier) {
                    file.error("expecting an identifier after 'new'", node.expression.pos + 1);
                }
                if (node.typeArguments !== undefined) {
                    try {
                        for (var _f = __values(node.typeArguments), _g = _f.next(); !_g.done; _g = _f.next()) {
                            var arg = _g.value;
                            lintType(file, arg);
                        }
                    }
                    catch (e_18_1) { e_18 = { error: e_18_1 }; }
                    finally {
                        try {
                            if (_g && !_g.done && (_a = _f.return)) _a.call(_f);
                        }
                        finally { if (e_18) throw e_18.error; }
                    }
                }
                if (node.arguments !== undefined) {
                    try {
                        for (var _h = __values(node.arguments), _j = _h.next(); !_j.done; _j = _h.next()) {
                            var arg = _j.value;
                            lintExpression(file, arg, isConstructor);
                        }
                    }
                    catch (e_19_1) { e_19 = { error: e_19_1 }; }
                    finally {
                        try {
                            if (_j && !_j.done && (_b = _h.return)) _b.call(_h);
                        }
                        finally { if (e_19) throw e_19.error; }
                    }
                }
            }
            break;
        case typescript_1.SyntaxKind.CallExpression:
            if ((0, types_1.assertCallExpr)(node)) {
                lintCallExpr(file, node, isConstructor);
            }
            break;
        case typescript_1.SyntaxKind.TemplateExpression:
            if ((0, types_1.assertTemplateExpr)(node)) {
                try {
                    for (var _k = __values(node.templateSpans), _l = _k.next(); !_l.done; _l = _k.next()) {
                        var span = _l.value;
                        lintExpression(file, span.expression, isConstructor);
                    }
                }
                catch (e_20_1) { e_20 = { error: e_20_1 }; }
                finally {
                    try {
                        if (_l && !_l.done && (_c = _k.return)) _c.call(_k);
                    }
                    finally { if (e_20) throw e_20.error; }
                }
            }
            break;
        case typescript_1.SyntaxKind.ArrayLiteralExpression:
            if ((0, types_1.assertArrayLiteralExpr)(node)) {
                try {
                    for (var _m = __values(node.elements), _o = _m.next(); !_o.done; _o = _m.next()) {
                        var elem = _o.value;
                        lintExpression(file, elem, isConstructor);
                    }
                }
                catch (e_21_1) { e_21 = { error: e_21_1 }; }
                finally {
                    try {
                        if (_o && !_o.done && (_d = _m.return)) _d.call(_m);
                    }
                    finally { if (e_21) throw e_21.error; }
                }
            }
            break;
        case typescript_1.SyntaxKind.ObjectLiteralExpression:
            if ((0, types_1.assertObjectLiteralExpr)(node)) {
                try {
                    for (var _p = __values(node.properties), _q = _p.next(); !_q.done; _q = _p.next()) {
                        var elem = _q.value;
                        lintObjectLiteralProperty(file, elem, isConstructor);
                    }
                }
                catch (e_22_1) { e_22 = { error: e_22_1 }; }
                finally {
                    try {
                        if (_q && !_q.done && (_e = _p.return)) _e.call(_p);
                    }
                    finally { if (e_22) throw e_22.error; }
                }
            }
            break;
        case typescript_1.SyntaxKind.SpreadElement:
            file.error("spread expressions are not allowed", node.pos + 1);
            break;
        case typescript_1.SyntaxKind.JsxElement:
            if ((0, types_1.assertJsxElem)(node)) {
                lintJsxElem(file, node, isConstructor);
            }
            break;
        case typescript_1.SyntaxKind.JsxSelfClosingElement:
            if ((0, types_1.assertJsxSelfClosingElem)(node)) {
                lintJsxSelfClosingElem(file, node, isConstructor);
            }
            break;
        case typescript_1.SyntaxKind.JsxFragment:
            if ((0, types_1.assertJsxFragment)(node)) {
                lintJsxFragment(file, node, isConstructor);
            }
            break;
        case typescript_1.SyntaxKind.ArrowFunction:
            if ((0, types_1.assertArrowFunction)(node)) {
                if (node.body.kind === typescript_1.SyntaxKind.Block && (0, types_1.assertBlock)(node.body)) {
                    file.error("only simple function expressions are allowed (not blocks)", node.parameters.pos, node.body.end);
                }
                else {
                    var text = node.getFullText(file.root).trim();
                    if (text.indexOf("\n") !== -1 || text.length > 100) {
                        file.error("only simple function expressions are allowed", node.parameters.pos, node.body.end);
                    }
                    else {
                        lintExpression(file, node.body, isConstructor);
                    }
                }
            }
            break;
        case typescript_1.SyntaxKind.FunctionExpression:
            if (!isTestFile(file)) {
                file.error("function expressions are not allowed here", node.pos, node.end);
            }
            else {
                if ((0, types_1.assertFunctionExpression)(node)) {
                    lintBlock(file, node.body, isConstructor);
                }
            }
            break;
        case typescript_1.SyntaxKind.NonNullExpression:
            if ((0, types_1.assertNonNullExpr)(node)) {
                file.error("non-null assertions are not allowed", node.end);
            }
            break;
        case typescript_1.SyntaxKind.ClassExpression:
            file.error("class expressions are not allowed", node.pos + 1);
            break;
        case typescript_1.SyntaxKind.AsExpression:
            file.error("type casts are not allowed", node.pos, node.end);
            break;
        case typescript_1.SyntaxKind.DeleteExpression:
            file.error("delete operator is not allowed", node.pos, node.end);
            break;
        case typescript_1.SyntaxKind.VoidExpression:
            file.error("void expression is not allowed", node.pos, node.end);
            break;
        case typescript_1.SyntaxKind.AwaitExpression:
            file.error("await expression is not allowed", node.pos, node.end);
            break;
        case typescript_1.SyntaxKind.YieldExpression:
            file.error("yield expression is not allowed", node.pos, node.end);
            break;
        case typescript_1.SyntaxKind.FirstTemplateToken:
            file.error("not a template string literal", node.pos, node.end);
            break;
        default:
            file.error("unexpected expression type: ".concat(typescript_1.SyntaxKind[node.kind]), node.pos, node.end);
            break;
    }
};
// Operators that perform assignments
var assignmentOperators = new Set([
    typescript_1.SyntaxKind.EqualsToken, typescript_1.SyntaxKind.PlusEqualsToken,
    typescript_1.SyntaxKind.MinusEqualsToken, typescript_1.SyntaxKind.AsteriskEqualsToken,
    typescript_1.SyntaxKind.AsteriskAsteriskEqualsToken, typescript_1.SyntaxKind.SlashEqualsToken,
    typescript_1.SyntaxKind.PercentEqualsToken, typescript_1.SyntaxKind.LessThanLessThanEqualsToken,
    typescript_1.SyntaxKind.GreaterThanGreaterThanEqualsToken,
    typescript_1.SyntaxKind.GreaterThanGreaterThanGreaterThanEqualsToken,
    typescript_1.SyntaxKind.AmpersandEqualsToken, typescript_1.SyntaxKind.BarEqualsToken,
    typescript_1.SyntaxKind.BarBarEqualsToken, typescript_1.SyntaxKind.AmpersandAmpersandEqualsToken,
    typescript_1.SyntaxKind.QuestionQuestionEqualsToken, typescript_1.SyntaxKind.CaretEqualsToken
]);
// Operators that perform boolean operations.
var booleanOperators = new Set([
    typescript_1.SyntaxKind.AmpersandAmpersandToken, typescript_1.SyntaxKind.BarBarToken
]);
// Operators that could be the root of a JSX subexpression.
var jsxOperators = new Set([
    typescript_1.SyntaxKind.JsxElement, typescript_1.SyntaxKind.JsxSelfClosingElement,
    typescript_1.SyntaxKind.JsxFragment
]);
// Perform linting of a binary expression.
var lintBinaryExpr = function (file, node, isConstructor) {
    lintExpression(file, node.left, isConstructor);
    lintExpression(file, node.right, isConstructor);
    if (booleanOperators.has(node.operatorToken.kind)) {
        if (jsxOperators.has(node.left.kind) || jsxOperators.has(node.right.kind)) {
            file.error("logical operators are not allowed on JSX elements", node.left.end, node.operatorToken.end);
        }
    }
    if (assignmentOperators.has(node.operatorToken.kind)) {
        if (file.allowMutation) {
            // okay
        }
        else if (isConstructor &&
            node.left.kind === typescript_1.SyntaxKind.PropertyAccessExpression &&
            (0, types_1.assertPropertyAccessExpr)(node.left) &&
            node.left.expression.kind === typescript_1.SyntaxKind.ThisKeyword) {
            // okay
        }
        else {
            file.error("mutation is not allowed", node.left.end, node.right.pos);
        }
    }
};
// Perform linting of a call expression.
var lintCallExpr = function (file, node, isConstructor) {
    var e_23, _a, e_24, _b;
    lintExpression(file, node.expression, isConstructor);
    if (node.typeArguments !== undefined) {
        try {
            for (var _c = __values(node.typeArguments), _d = _c.next(); !_d.done; _d = _c.next()) {
                var arg = _d.value;
                lintType(file, arg);
            }
        }
        catch (e_23_1) { e_23 = { error: e_23_1 }; }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
            }
            finally { if (e_23) throw e_23.error; }
        }
    }
    if (node.expression.kind === typescript_1.SyntaxKind.Identifier &&
        (0, types_1.assertIdentifier)(node.expression)) {
        var funcName = node.expression.text;
        if (funcName === "fetch") {
            lintFetch(file, node);
        }
    }
    // Changing conventions to use .json and .send
    //  } else if (node.expression.kind === SyntaxKind.PropertyAccessExpression &&
    //             assertPropertyAccessExpr(node.expression) &&
    //             node.expression.name.kind === SyntaxKind.Identifier &&
    //             assertIdentifier(node.expression.name)) {
    //    const methodName = node.expression.name.text;
    //    if ((methodName === "json" || methodName === "text") &&
    //        node.arguments.length === 1) {
    //      file.error(`call to .${methodName} is not allowed`, node.expression.name.end - 1);
    //    }
    //  }
    if (node.arguments !== undefined) {
        try {
            for (var _e = __values(node.arguments), _f = _e.next(); !_f.done; _f = _e.next()) {
                var arg = _f.value;
                lintExpression(file, arg, isConstructor);
            }
        }
        catch (e_24_1) { e_24 = { error: e_24_1 }; }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
            }
            finally { if (e_24) throw e_24.error; }
        }
    }
};
// Performs linting of a property within an object literal expression.
var lintObjectLiteralProperty = function (file, elem, isConstructor) {
    switch (elem.kind) {
        case typescript_1.SyntaxKind.PropertyAssignment:
            if ((0, types_1.assertPropertyAssignment)(elem)) {
                lintExpression(file, elem.initializer, isConstructor);
            }
            break;
        case typescript_1.SyntaxKind.ShorthandPropertyAssignment:
            // OK
            break;
        case typescript_1.SyntaxKind.SpreadAssignment:
            file.error("spread expressions are not allowed", elem.pos, elem.end);
            break;
        case typescript_1.SyntaxKind.MethodDeclaration:
            file.error("method declarations are not allowed", elem.pos, elem.end);
            break;
        default:
            file.error("unexpected property type: ".concat(typescript_1.SyntaxKind[elem.kind]), elem.pos, elem.end);
            break;
    }
};
// Performs linting of a JSX element.
var lintJsxElem = function (file, node, isConstructor) {
    var e_25, _a;
    lintJsxAttributes(file, node.openingElement.attributes, isConstructor);
    try {
        for (var _b = __values(node.children), _c = _b.next(); !_c.done; _c = _b.next()) {
            var child = _c.value;
            lintJsxChild(file, child, isConstructor);
        }
    }
    catch (e_25_1) { e_25 = { error: e_25_1 }; }
    finally {
        try {
            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
        }
        finally { if (e_25) throw e_25.error; }
    }
};
// Performs linting of a JSX self-closing element.
var lintJsxSelfClosingElem = function (file, node, isConstructor) {
    lintJsxAttributes(file, node.attributes, isConstructor);
};
// Performs linting of a JSX element.
var lintJsxAttributes = function (file, attributes, isConstructor) {
    var e_26, _a;
    try {
        for (var _b = __values(attributes.properties), _c = _b.next(); !_c.done; _c = _b.next()) {
            var attr = _c.value;
            if (attr.kind !== typescript_1.SyntaxKind.JsxAttribute) {
                file.error("spread attributes are not allowed", attr.pos, attr.end);
            }
            else if ((0, types_1.assertJsxAttribute)(attr)) {
                if (attr.initializer !== undefined) {
                    if (attr.initializer.kind === typescript_1.SyntaxKind.StringLiteral) {
                        // OK
                    }
                    else if (attr.initializer.kind === typescript_1.SyntaxKind.JsxExpression &&
                        (0, types_1.assertJsxExpr)(attr.initializer)) {
                        var expr = attr.initializer.expression;
                        if (expr !== undefined) {
                            lintExpression(file, expr, isConstructor);
                        }
                    }
                    else {
                        file.error("unexpected attribute type: ".concat(typescript_1.SyntaxKind[attr.initializer.kind]), attr.pos, attr.end);
                    }
                }
            }
        }
    }
    catch (e_26_1) { e_26 = { error: e_26_1 }; }
    finally {
        try {
            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
        }
        finally { if (e_26) throw e_26.error; }
    }
};
// Performs linting of a JSX fragment.
var lintJsxFragment = function (file, node, isConstructor) {
    var e_27, _a;
    try {
        for (var _b = __values(node.children), _c = _b.next(); !_c.done; _c = _b.next()) {
            var child = _c.value;
            lintJsxChild(file, child, isConstructor);
        }
    }
    catch (e_27_1) { e_27 = { error: e_27_1 }; }
    finally {
        try {
            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
        }
        finally { if (e_27) throw e_27.error; }
    }
};
// Performs linting of a JSX child.
var lintJsxChild = function (file, child, isConstructor) {
    switch (child.kind) {
        case typescript_1.SyntaxKind.JsxText:
            // OK
            break;
        case typescript_1.SyntaxKind.JsxExpression:
            if ((0, types_1.assertJsxExpr)(child)) {
                if (child.expression !== undefined) {
                    lintExpression(file, child.expression, isConstructor);
                }
            }
            break;
        case typescript_1.SyntaxKind.JsxElement:
            lintJsxElem(file, child, isConstructor);
            break;
        case typescript_1.SyntaxKind.JsxSelfClosingElement:
            lintJsxSelfClosingElem(file, child, isConstructor);
            break;
        case typescript_1.SyntaxKind.JsxFragment:
            lintJsxFragment(file, child, isConstructor);
            break;
        default:
            throw new Error("impossible");
    }
};
// Performs linting of a call to fetch.
var lintFetch = function (file, node) {
    if (node.arguments.length === 1) {
        // OK
    }
    else if (node.arguments.length === 2) {
        var arg = node.arguments[1];
        if (arg.kind !== typescript_1.SyntaxKind.ObjectLiteralExpression) {
            file.error("second argument to fetch should be a record", arg.pos, arg.end);
        }
        else if ((0, types_1.assertObjectLiteralExpr)(arg)) {
            var fields = getFields(arg);
            var method = fields.get("method");
            if (method === undefined) {
                file.error("missing 'method' field in the call to fetch", arg.pos, arg.end);
            }
            else if (method.kind !== typescript_1.SyntaxKind.StringLiteral ||
                (0, types_1.assertStringLiteral)(method) && (method.text !== "POST")) {
                file.error("'method' is only required for POST requests", method.pos, method.end);
            }
            var body = fields.get("body");
            if (body === undefined) {
                file.error("missing 'body' field in the call to fetch", arg.pos, arg.end);
            }
            var headers = fields.get("headers");
            if (headers === undefined) {
                file.error("missing 'headers' field in the call to fetch", arg.pos, arg.end);
            }
            else if (headers.kind === typescript_1.SyntaxKind.ObjectLiteralExpression &&
                (0, types_1.assertObjectLiteralExpr)(headers)) {
                var headerFields = getFields(headers);
                var contentType = headerFields.get("Content-Type");
                if (contentType === undefined) {
                    file.error("missing 'Content-Type' header in the call to fetch", arg.pos, arg.end);
                }
                else if (contentType.kind !== typescript_1.SyntaxKind.StringLiteral ||
                    (0, types_1.assertStringLiteral)(contentType) && (contentType.text !== "application/json")) {
                    file.error("'Content-Type' header should be 'application/json'", arg.pos, arg.end);
                }
            }
            else {
                file.error("'headers' should be an object literal", arg.pos, arg.end);
            }
        }
    }
    else {
        file.error("expecting one or two arguments to fetch", node.arguments.pos, node.arguments.end);
    }
};
// Returns a map from constant field names to their value expressions in an
// object literal expression (non-constant fields are ignored).
var getFields = function (expr) {
    var e_28, _a;
    var fields = new Map();
    try {
        for (var _b = __values(expr.properties), _c = _b.next(); !_c.done; _c = _b.next()) {
            var prop = _c.value;
            if (prop.kind === typescript_1.SyntaxKind.PropertyAssignment &&
                (0, types_1.assertPropertyAssignment)(prop)) {
                var name_2 = getName(prop.name);
                if (name_2 !== undefined) {
                    fields.set(name_2, prop.initializer);
                }
            }
        }
    }
    catch (e_28_1) { e_28 = { error: e_28_1 }; }
    finally {
        try {
            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
        }
        finally { if (e_28) throw e_28.error; }
    }
    return fields;
};
// Returns the name of a property or undefined if not unedrstood.
var getName = function (name) {
    if (name.kind === typescript_1.SyntaxKind.Identifier && (0, types_1.assertIdentifier)(name)) {
        return name.text;
    }
    else if (name.kind === typescript_1.SyntaxKind.StringLiteral && (0, types_1.assertStringLiteral)(name)) {
        return name.text;
    }
    else {
        return undefined;
    }
};
